homeassistant/helpers/config_validation.py

```diff
@@ -441,6 +441,15 @@ def url(value: Any) -> str:
     raise vol.Invalid('invalid url')
 
 
+def ip_address(value: Any) -> str:
+    """Validate an IP address."""
+    value = str(value).strip()
+    pattern = re.compile(r'^(\d{1,3}.){3}(\d{1,3})$')
```

Comment:
Won't the dot `.` match all characters here? I think currently this accepts `0&0€0@0`.

Also, there's not need to `compile` here as it will be evaluated in every call anyway. Just call `re.match(r'...', value)`.

It might also be good to validate that each part of the IP address is in the range 0-255.